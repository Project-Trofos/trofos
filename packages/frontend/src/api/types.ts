export type Project = {
  id: number;
  pname: string;
  pkey: string | null;
  description: string | null;
  course_id: number | null;
  public: boolean;
  created_at: string;
  course?: Course;
  is_archive: boolean | null;
};

export type Course = {
  id: number;
  code: string;
  startYear: number;
  startSem: number;
  endYear: number;
  endSem: number;
  cname: string;
  description: string | null;
  public: boolean;
  created_at: string;
  shadow_course: boolean;
  is_archive: boolean | null;
};

export type User = {
  user_email: string;
  user_display_name: string;
  user_id: number;
  projects: Project[];
  courses: CourseRoles[];
  basicRoles: BasicRoles[];
};

export type BasicRoles = {
  user_email: string;
  role_id: number;
};

export type CourseRoles = {
  id: number;
  user_email: string;
  role_id: number;
  course_id: number;
};

export type CreateUserRequest = {
  userEmail: string;
  newPassword: string;
};

export type RegisterUser = {
  userEmail: string;
  newPassword: string;
  userDisplayName: string;
};

export type UserCourseRoleRequest = {
  id: number;
  userRole: number;
  userId: number;
};

export type Role = {
  role_name: string;
  id: number;
};

export type Action = {
  action: string;
};

export type ActionOnRole = {
  id: number;
  action: string;
};

export type ActionsOnRoles = {
  id: number;
  role_name: string;
  actions: Action[];
};

export type ProjectData = Project & {
  course: Course | null;
  users: UserData[];
  sprints: {
    id: number;
    name: string;
  }[];
  backlogStatuses: Omit<BacklogStatusData, 'project_id'>[];
  telegramChannelLink: string;
  owner_id: number | null;
};

export type UserData = {
  user: Pick<User, 'user_id' | 'user_email' | 'user_display_name' | 'courses'>;
};

export type ScrumBoardUserData = {
  user: {
    user_id: number | null;
    user_display_name: string;
    user_email: string;
  };
};

export type Milestone = {
  id: number;
  name: string;
  start_date: string;
  deadline: string;
  course_id: number;
  created_at: string;
};

export type Announcement = {
  id: number;
  title: string;
  content: string;
  user_id: number;
  created_at: string;
  updated_at?: string;
};

// CourseData type returned by BE api. We perform transformResponse to convert the data into type CourseData
export type CourseDataResponse = Omit<CourseData, 'users'> & {
  courseRoles: UserData[];
};

export type CourseData = Course & {
  milestones: Milestone[];
  announcements: Announcement[];
  users: UserData[];
};

export type UserOnRolesOnCourse = {
  id: number;
  user_id: number;
  role_id: number;
  course_id: number;
  role: {
    id: number;
    role_name: string;
  };
};

export type BacklogStatusData = {
  name: string;
  type: 'todo' | 'in_progress' | 'done';
  project_id: number;
  order: number;
};

export type Comment = {
  comment_id: number;
  backlog_id: number;
  project_id: number;
  commenter_id: number;
  content: string;
  created_at: string;
  updated_at: string | null;
  commenter: {
    created_at: string;
    project_id: string;
    user: {
      user_id: number;
      user_email: string;
      user_display_name: string;
    };
  };
};

export type CommentFieldsType = {
  projectId: number;
  backlogId: number;
  commenterId: number;
  content: string;
};

export enum BacklogStatus {
  TODO = 'To do',
  DONE = 'Done',
}

export type BacklogPriority = 'very_high' | 'high' | 'medium' | 'low' | 'very_low' | null;

export type Backlog = {
  backlog_id: number;
  summary: string;
  type: 'story' | 'task' | 'bug';
  priority: BacklogPriority;
  reporter_id: number;
  assignee_id: number | null;
  sprint_id: number | null;
  points: number | null;
  description: string | null;
  project_id: number;
  status: BacklogStatus | string;
  assignee: {
    created_at: string;
    project_id: number;
    user_id: number;
    user: {
      user_email: string;
      user_display_name: string;
    };
  } | null;
  epic_id?: number | null;
  epic?: Epic | null;
};

export type Epic = {
  epic_id: number;
  project_id: number;
  name: string;
  description: string | null;
};

export type DefaultBacklog = {
  backlog_id?: number;
  summary?: string;
  type?: 'story' | 'task' | 'bug';
  priority?: BacklogPriority;
  reporter_id?: number;
  assignee_id?: number | null;
  sprint_id?: number | null;
  points?: number | null;
  description?: string | null;
  project_id?: number;
  status?: BacklogStatus | string;
  assignee?: {
    created_at: string;
    project_id: number;
    user_id: number;
    user: {
      user_email: string;
      user_display_name: string;
    };
  } | null;
  epic_id?: number | null;
};

export type BacklogUpdatePayload = {
  projectId: number;
  backlogId: number;
  srcSprintId?: number | null;
  fieldToUpdate: Partial<Backlog>;
};

export enum BacklogHistoryType {
  CREATE = 'create',
  UPDATE = 'update',
  DELETE = 'delete',
}

export type BacklogHistory = {
  history_type: BacklogHistoryType;
  date: string;
} & Omit<Backlog, 'assignee' | 'summary' | 'description'>;

export type UpdateUserRolePayload = {
  userId: number;
  newRoleId: number;
};

export type ProjectGitLink = {
  project_id: number;
  repo: string;
};

export type ProjectGitLinkData = {
  projectId: number;
  repoLink: string;
};

export type ProjectUserSettings = {
  project_id: number;
  user_id: number;
  email_notification: boolean;
};

export type Settings = {
  current_year: number;
  current_sem: number;
};

export type RetrospectiveVote = {
  id: number;
  retro_id: number;
  user_id: number;
  type: RetrospectiveVoteType;
};

export type Retrospective = {
  id: number;
  sprint_id: number;
  content: string;
  type: RetrospectiveType;
  score: number;
  votes: RetrospectiveVote[];
  is_action_taken: boolean;
};

export enum RetrospectiveType {
  POSITIVE = 'positive',
  NEGATIVE = 'negative',
  ACTION = 'action',
}

export enum RetrospectiveVoteType {
  UP = 'up',
  DOWN = 'down',
}

export type CourseImportCsvPayload = {
  courseId: number;
  payload: FormData;
};

export type Feedback = {
  id: number;
  user_id?: number;
  user: {
    user_email: string;
    user_display_name: string;
  };
  sprint_id: number;
  content: string;
  created_at: string;
  updated_at?: string;
};

export type OAuth2Payload = {
  code: string;
  state: string;
  callbackUrl: string;
};

export type Invite = {
  project_id: number;
  email: string;
  expiry_date: Date;
};

export type UserApiKey = {
  id: number;
  user_id: number;
  api_key: string;
  created_at: string;
  last_used: string | null;
  active: boolean;
};

export type UserGuideQueryResponse = {
  answer: string;
  links: Array<string>;
};

export type FeatureFlag = {
  feature_name: string;
  active: boolean;
};

export type ProjectAssignment = {
  id: number;
  targetProjectId: number;
  sourceProjectId: number;
};
