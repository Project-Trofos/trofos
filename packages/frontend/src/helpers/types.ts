export type PickRename<T, K extends keyof T, R extends PropertyKey> = Pick<Omit<T, K> & { [P in R]: T[K] }, R>;
