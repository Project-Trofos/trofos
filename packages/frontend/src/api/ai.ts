import trofosApiSlice from '.';
import { UserGuideQueryResponse, UserGuideRecommendation } from './types';

const extendedApi = trofosApiSlice.injectEndpoints({
  endpoints: (builder) => ({
    answerUserGuideQuery: builder.mutation<UserGuideQueryResponse, { query: string; isEnableMemory: boolean }>({
      query: ({ query, isEnableMemory }) => ({
        url: `ai/userGuideQuery`,
        credentials: 'include',
        method: 'POST',
        body: { query, isEnableMemory },
      }),
    }),
    recommendUserGuideSections: builder.mutation<UserGuideRecommendation[], void>({
      query: () => ({
        url: `ai/recommendUserGuide`,
        method: 'POST',
        credentials: 'include',
      }),
    }),
  }),
});

export const { useAnswerUserGuideQueryMutation, useRecommendUserGuideSectionsMutation } = extendedApi;
