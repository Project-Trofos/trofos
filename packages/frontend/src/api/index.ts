import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

// Define a service using a base URL and expected endpoints
const trofosApiSlice = createApi({
  reducerPath: 'trofosApi',
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NODE_ENV === 'test' ? 'http://localhost:3001/api' : '/api',
    prepareHeaders: (headers) => {
      headers.set('Cache-control', 'no-cache');
      return headers;
    },
  }),
  tagTypes: [
    'Project',
    'UserInfo',
    'Course',
    'Backlog',
    'BacklogHistory',
    'User',
    'Sprint',
    'ActionsOnRoles',
    'CourseRoles',
    'ProjectRoles',
    'Settings',
    'StandUpHeader',
    'StandUp',
    'Retrospective',
    'ProjectUserSettings',
    'Feedback',
  ],
  endpoints: () => ({}),
});

// Define a service using a base URL and expected endpoints
export const nusmodsApiSlice = createApi({
  reducerPath: 'nusmodsApi',
  baseQuery: fetchBaseQuery({ baseUrl: 'https://api.nusmods.com/v2/2022-2023' }),
  endpoints: () => ({}),
});

export default trofosApiSlice;
