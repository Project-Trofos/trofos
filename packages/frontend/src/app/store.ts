import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
