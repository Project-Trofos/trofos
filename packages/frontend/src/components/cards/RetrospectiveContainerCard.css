.retrospective-container-card {
  margin: 5px;
  width: 33%;
  display: flex;
  flex-flow: column;
}

.retrospective-container-card .ant-card-head-title {
  text-align: center;
}

.retrospective-container-card .ant-card-body {
  padding: 12px;
  flex: 1 1 auto;
  display: flex;
  flex-flow: column;
}

.retrospective-container-card-body {
  min-height: 2vh;
  overflow-y: auto;
}

.retrospective-container-card-footer {
  margin-top: auto;
}

.retrospective-container-form .ant-form-item {
  margin-bottom: 0;
}

.retrospective-container-form input {
  height: 44px;
}
