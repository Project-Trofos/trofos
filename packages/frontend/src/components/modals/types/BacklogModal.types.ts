export interface BacklogFormFields extends FormData {
  projectId: number;
}

export type BacklogSelect = {
  id: string;
  name: string;
};
