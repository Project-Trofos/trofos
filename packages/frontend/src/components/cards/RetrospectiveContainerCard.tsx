import React from 'react';
import { Card, Divider, Form, Input } from 'antd';
import { useGetRetrospectivesQuery } from '../../api/sprint';
import { useAddRetrospectiveMutation } from '../../api/socket/retrospectiveHooks';
import { RetrospectiveType } from '../../api/types';
import RetrospectiveContentCard from './RetrospectiveContentCard';
import './RetrospectiveContainerCard.css';

export default function RetrospectiveContainerCard(props: RetrospectiveContainerCardProps): JSX.Element {
  const { title, type, sprintId } = props;
  const [form] = Form.useForm();

  const { data: retrospectivesData } = useGetRetrospectivesQuery({ sprintId, type });
  const [addRetrospective] = useAddRetrospectiveMutation();

  const handleFormSubmit = async (formData: { content: string }) => {
    const { content } = formData;
    if (!content) return;

    try {
      await addRetrospective({ sprintId, content, type }).unwrap();
      form.resetFields();
    } catch (err) {
      console.error(err);
    }
  };

  return (
    <Card className="retrospective-container-card" title={title}>
      <div className="retrospective-container-card-body">
        {retrospectivesData?.map((retro) => (
          <RetrospectiveContentCard key={retro.id} retroEntry={retro} />
        ))}
      </div>
      <div className="retrospective-container-card-footer">
        <Divider />
        <Form className="retrospective-container-form" form={form} onFinish={handleFormSubmit}>
          <Form.Item name="content">
            <Input placeholder="Type something here..." />
          </Form.Item>
        </Form>
      </div>
    </Card>
  );
}

type RetrospectiveContainerCardProps = {
  title: string;
  type: RetrospectiveType;
  sprintId: number;
};
