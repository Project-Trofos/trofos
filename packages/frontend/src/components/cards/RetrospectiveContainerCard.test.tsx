import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import RetrospectiveContainerCard from './RetrospectiveContainerCard';
import store from '../../app/store';
import { RetrospectiveType } from '../../api/types';

describe('RetrospectiveContainerCard test', () => {
  const MOCK_SPRINT_ID = 1;
  const MOCK_TITLE = 'What went well?';
  const MOCK_TYPE = RetrospectiveType.POSITIVE;

  const { baseElement } = render(
    <Provider store={store}>
      <RetrospectiveContainerCard sprintId={MOCK_SPRINT_ID} title={MOCK_TITLE} type={MOCK_TYPE} />
    </Provider>,
  );

  it('renders retrospective card with correct details', () => {
    expect(screen.getByText(MOCK_TITLE)).toBeInTheDocument();

    // Compare with snapshot to ensure structure remains the same
    expect(baseElement).toMatchSnapshot();
  });
});
