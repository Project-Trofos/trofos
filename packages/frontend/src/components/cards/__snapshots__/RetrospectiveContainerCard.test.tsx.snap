// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RetrospectiveContainerCard test > renders retrospective card with correct details 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered retrospective-container-card"
    >
      <div
        class="ant-card-head"
      >
        <div
          class="ant-card-head-wrapper"
        >
          <div
            class="ant-card-head-title"
          >
            What went well?
          </div>
        </div>
      </div>
      <div
        class="ant-card-body"
      >
        <div
          class="retrospective-container-card-body"
        />
        <div
          class="retrospective-container-card-footer"
        >
          <div
            class="ant-divider ant-divider-horizontal"
            role="separator"
          />
          <form
            class="ant-form ant-form-horizontal retrospective-container-form"
          >
            <div
              class="ant-form-item"
            >
              <div
                class="ant-row ant-form-item-row"
              >
                <div
                  class="ant-col ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <input
                        class="ant-input"
                        id="content"
                        placeholder="Type something here..."
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</body>
`;
